rfcs/20190502-preprocessing-layers.md

```diff
@@ -0,0 +1,578 @@
+
+
+### Handling of async prefetching
+
+Some preprocessing ops are CPU-only and benefit from being executed asynchronously on the accelerator host (as opposed to the accelerator itself, e.g. GPU or TPU),
+with a batch of data being prepocessed on the host while the previous batch is being processed by the accelerator. This pattern is known as "async prefetching".
+
+This is normally done as part of a tf.data pipeline. The current proposal implies moving some of that preprocessing to inside the model itself, which is normally
+executed end-to-end on an accelerator.
+
+This means that we need a way to lift the preprocessing part of the model in a tf.data pipeline during model training. In `fit`, we can do this automatically.
+In custom training loops, we will expect the user to do it manually (see subsection "Custom training loops").
+
+We propose the addition of two new methods on the `Model` class:
+
+```python
+def get_preprocessing_stage(self):
+    """Retrieves the preprocessing part of the model.
+
+    This is the part of the model that should be executed asynchronously
+    on the device host during training.
+
+    Returns:
+        Instance of `PreprocessingLayer` or `PreprocessingStage`.
+        May be None if the model does not start with preprocessing layers.
+    """
+    pass
+
+def get_main_stage(self):
+    """Retrieves the main processing part of the model.
+
+    This is the part of the model that should be executed
+    on the accelator device.
+
+    Returns:
+        Model instance.
+    """
+```
+
+Thus, for any model that starts with preprocessing layers, the following:
+
+```python
+outputs = model(inputs)
+```
+
+is functionnally equivalent to:
+
+```python
+preprocessed_inputs = model.get_preprocessing_stage()(inputs)
+outputs = model.get_main_stage()(preprocessed_inputs)
+```
+
+
+#### Examples:
+
+Sequential model with a preprocessing layer:
+
+```python
+vectorization = keras.layers.TextVectorization()
+vectorization.adapt(data_sample)
+
+model = keras.Sequential([
+    vectorization,
+    keras.layers.Dense(10, activation='softmax'),
+])
+
+# This is the `vectorization` layer.
+preproc_stage = model.get_preprocessing_stage()
+# model containing the `Dense` layer only.
+main_stage = model.get_main_stage()
+```
+
+Functional model with 2 branches, each with a preprocessing layer:
+
+```python
+normalization_a = layers.Normalization()
+normalization_b = layers.Normalization()
+normalization_a.adapt(data_a)
+normalization_b.adapt(data_b)
+
+input_a = Input(shape_a)
+input_b = Input(shape_b)
+normed_a = normalization_a(input_a)
+normed_b = normalization_b(input_b)
+a = layers.Dense(32)(normed_a)
+b = layers.Dense(32)(normed_b)
+c = layers.concatenate([a, b])
+outputs = layers.Dense(1, activation='sigmoid')(c)
+
+model = Model([input_a, input_b], outputs)
+
+# `PreprocessingStage` instance
+# mapping `[input_a, input_b]` to `[normed_a, normed_b]`
```

Comment:
`Normalization` can benefit from being done on the accelerator. 
* Without the graph optimization, the computing time of the `Normalization` is negligible compared to, let's say, a `Dense` or `Convolutional` layer, because it's an elementwise operation.
* With a graph optimization, it can be folded into the first layer, since first layers are often linear operations. Making the preprocessing time null. This may not be done with async prefetching. I might be wrong about this second point, I guess it depends how much the graph is optimized behind the scenes.

Reply:
This is a good point in general: the hypothesis that a preprocessing layer should be moved to async prefetching isn't always true. We'll have to keep this in mind.

In the case of `Normalization`, I would think it is cheap enough that it wouldn't matter whether we run it on the accelerator or on the host.